Feature/task 3/impl

Pavel-Laukhin opened the pull request:
Задание номер 3

Pavel-Laukhin committed:
Changed scroll view behavior and keyboard observers in HelloViewController. Now the content of the screen scrolls automatically after the editing starts and you can scroll the content as you want.

Pavel-Laukhin committed:
1. Added ActivityIndicatorViewController.
2. Replaced keyboard observers in HelloViewController to Nikita's. They have be overhauled for self-scrolling.
3. Added ResultTableViewController and its ResultTableViewCell.
4. Rebuild QueryEngine for ability to serialize JSON and to show table with search results.
5. Added Repo structure.
6. Made some architecture changes.

Pavel-Laukhin committed:
Changed keyboard observers.

nskazakov commented on GitHubClient/Controllers/HelloViewController.swift:
```diff
@@ -226,24 +232,68 @@ class HelloViewController: UIViewController {
+            // Считаем размер, на который нужно подвинуть содержимое скролл вью, в зависимости от того, пересекает ли клавиатура фрейм сегментед контрола (чтобы сегментед контролл был видимым).
+            // Контент должен отъехать на разницу координат верхней точки клавиатуры и нижней точки сегментед контрола. Добавим еще отступ 8 поинтов, чтобы сегментед контрол не прилипал к клавиатуре.
+            let keyboardOffset = (convertedKeyboardFrame.intersects(segmentedControl.frame)) ? segmentedControl.frame.maxY - convertedKeyboardFrame.minY + 8 : 0
+            UIView.animate(withDuration: 0.2) {
+//                self.scrollView.contentOffset.y = keyboardOffset
+            }
```
можно удалить неиспользуемый код

nskazakov commented on GitHubClient/Controllers/ResultTableViewCell.swift:
```diff
@@ -0,0 +1,112 @@
+}
+
+
+extension UIView {
+    
+    func toAutoLayout() {
+        translatesAutoresizingMaskIntoConstraints = false
+    }
+    
+}
```
лучше вынести в отдельный файл

Pavel-Laukhin replied:
Вынес, спасибо!

nskazakov commented on GitHubClient/Controllers/ResultTableViewController.swift:
```diff
@@ -0,0 +1,104 @@
+
+extension ResultTableViewController: UITableViewDelegate {
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: true)
+    }
```
можно у таблицы определить selectionStyle на none и тогда не придется в делегате в этом методе определять

Pavel-Laukhin replied:
Исправил, спасибо. Правда не selectionStyle, а allowsSelection, и не на none, а на false. Что-то я не нашел там никакого selectionStyle ((

nskazakov commented on GitHubClient/Models/Repo.swift:
```diff
@@ -0,0 +1,36 @@
+
+import UIKit
+
+struct Repo {
```
почему не Codable?

Pavel-Laukhin replied:
Потому что не разобрался в том, как он работает )) Теперь разобрался, окей ))

Pavel-Laukhin committed:
1. Deleted unused code in HelloViewController
2. "toAutoLayout" method in UIView's extension has been relocated to separate file.
3. Deleted deselecting rows from ResultTableViewController's extension. Instead of that there "allowsSelection" property of the table has been set to false.

Pavel-Laukhin committed:
1. The Repo structure became nested Codable structure in the Repos structure (also Codable).
2. Serialisation of JSON has been removed by decoding with JSONDecoder instance in QueryEngine.
3. Because of new nested type of the Repo structure, there are several edits in ResultTableViewCell and ResultTableViewController: [Repos.Repo] instead of [Repo].

Pavel-Laukhin committed:
A little code cleaning )
